source/common/network/udp_listener_impl.cc

```diff
@@ -68,50 +68,23 @@ void UdpListenerImpl::onSocketEvent(short flags) {
 
 void UdpListenerImpl::handleReadCallback() {
   ENVOY_UDP_LOG(trace, "handleReadCallback");
-  do {
-    uint32_t old_packets_dropped = packets_dropped_;
-    MonotonicTime receive_time = time_source_.monotonicTime();
-    Api::IoCallUint64Result result =
-        Utility::readFromSocket(socket_, *this, receive_time, &packets_dropped_);
-
-    if (!result.ok()) {
-      // No more to read or encountered a system error.
-      if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
-        ENVOY_UDP_LOG(error, "recvmsg result {}: {}", static_cast<int>(result.err_->getErrorCode()),
-                      result.err_->getErrorDetails());
-        cb_.onReceiveError(UdpListenerCallbacks::ErrorCode::SyscallError,
-                           result.err_->getErrorCode());
-      }
-      // Stop reading.
-      return;
-    }
-
-    if (result.rc_ == 0) {
-      // TODO(conqerAtapple): Is zero length packet interesting? If so add stats
-      // for it. Otherwise remove the warning log below.
-      ENVOY_UDP_LOG(trace, "received 0-length packet");
-    }
-
-    if (packets_dropped_ != old_packets_dropped) {
-      // The kernel tracks SO_RXQ_OVFL as a uint32 which can overflow to a smaller
-      // value. So as long as this count differs from previously recorded value,
-      // more packets are dropped by kernel.
-      uint32_t delta = (packets_dropped_ > old_packets_dropped)
-                           ? (packets_dropped_ - old_packets_dropped)
-                           : (packets_dropped_ +
-                              (std::numeric_limits<uint32_t>::max() - old_packets_dropped) + 1);
-      // TODO(danzh) add stats for this.
-      ENVOY_UDP_LOG(debug, "Kernel dropped {} more packets. Consider increase receive buffer size.",
-                    delta);
-    }
-  } while (true);
+  const Api::IoCallUint64Result result = Utility::readAllDatagramsFromSocket(
+      socket_.ioHandle(), *socket_.localAddress(), *this, time_source_, packets_dropped_);
+  // We should always get a failed result, even if no further data.
```

Comment:
Could maybe provide a little more detail on this comment, e.g. "readAllDatagramsFromSocket should/will always read until we get an error, though that error may simply be IoErrorCode::Again".